Mathlib/RingTheory/LocalProperties/Flat.lean

```diff
@@ -0,0 +1,214 @@
+    Localization S ⊗[R] M ≃ₗ[Localization S] LocalizedModule S M :=
+  (IsLocalizedModule.isBaseChange S (Localization S) (mkLinearMap S M)).equiv
+
+private noncomputable def eqv1 := (TensorProduct.assoc R (Localization S) M N)
+  ≪≫ₗ ((tensor_eqv_local (M ⊗[R] N) S).restrictScalars R)
+
+private noncomputable def eqv2 := TensorProduct.congr
+  ((TensorProduct.rid (Localization S) (Localization S ⊗[R] M)).restrictScalars R) (refl R N)
+
+private noncomputable def eqv3 := (AlgebraTensorModule.assoc R (Localization S) (Localization S)
+  (Localization S ⊗[R] M) (Localization S) N).symm
+
+private noncomputable def eqv4 := TensorProduct.congr
+  (tensor_eqv_local M S).symm (tensor_eqv_local N S).symm
+
+private noncomputable def eqv' := ((eqv4 M N S).restrictScalars R)
+  ≪≫ₗ (((eqv3 M N S).restrictScalars R) ≪≫ₗ ((eqv2 M N S) ≪≫ₗ (eqv1 M N S)))
+
+private noncomputable def lmap := (eqv' M N S).extendScalarsOfIsLocalization S (Localization S)
+
+private noncomputable def rmap := (eqv' M N S).symm.extendScalarsOfIsLocalization S (Localization S)
+
+private noncomputable def eqv := ofLinear (lmap M N S) (rmap M N S) (re₂₁ := RingHomInvPair.ids)
+(re₁₂ := RingHomInvPair.ids) (by
+  unfold lmap rmap
+  ext x
+  simp only [coe_comp, Function.comp_apply, extendScalarsOfIsLocalization_apply',
+    LinearEquiv.coe_coe, apply_symm_apply, id_coe, id_eq]) (by
+  unfold lmap rmap
+  ext x
+  simp only [AlgebraTensorModule.curry_apply, restrictScalars_comp, curry_apply, coe_comp,
+    LinearMap.coe_restrictScalars, Function.comp_apply, extendScalarsOfIsLocalization_apply',
+    LinearEquiv.coe_coe, symm_apply_apply, id_coe, id_eq])
+
+lemma tensor_eqv_local_apply (m : M) (sm : S) (r : R) :
+    (tensor_eqv_local M S) (Localization.mk r sm ⊗ₜ[R] m) = (LocalizedModule.mk (r • m) sm) := by
+  rw [tensor_eqv_local, equiv_tmul, mkLinearMap_apply, mk_smul_mk, mul_one]
+
+lemma tensor_eqv_local_symm_apply (m : M) (sm : S) :
+    (tensor_eqv_local M S).symm (LocalizedModule.mk m sm) = Localization.mk 1 sm ⊗ₜ[R] m :=
+  (symm_apply_eq _).mpr <| by
+  have := ((one_smul R m) ▸ (tensor_eqv_local_apply _ _ m sm 1).symm); exact this
+
+lemma eqv'_mk_apply (m : M) (n : N) (sm sn : S) :
+    (eqv' M N S) (mk m sm ⊗ₜ[Localization S] mk n sn) = mk (m ⊗ₜ[R] n) (sm * sn) := by
+  unfold eqv'
+  simp only [trans_apply, LinearEquiv.restrictScalars_apply, eqv4, congr_tmul,
+    tensor_eqv_local_symm_apply, eqv3, AlgebraTensorModule.assoc_symm_tmul, eqv2, refl_apply,
+    rid_tmul, eqv1, smul_tmul', assoc_tmul, smul_eq_mul, Localization.mk_mul, one_mul, mul_comm,
+    tensor_eqv_local_apply, one_smul]
+
+end Tensor
+
+section flatlocal
+variable {R : Type*} (M N : Type*) [CommRing R] (S : Submonoid R) [AddCommGroup M] [Module R M]
+  [AddCommGroup N] [Module R N]
+
+theorem Flat_of_localization_maximal (h : ∀ (J : Ideal R) (_ : J.IsMaximal),
+    Module.Flat (Localization.AtPrime J) (LocalizedModule J.primeCompl M)) : Module.Flat R M := by
```

Comment:
It's possible to shorten these two proofs considerably and get rid of the auxiliary defs by proving equality between the two maps $(M \otimes I)_J \to M_J$ and $(M \otimes I)_J \cong R_J \otimes (M \otimes I) \cong (R_J \otimes M)\otimes I \cong M_J \otimes I \to M_J$ (where all tensor products are over R). Please refer to [this commit](https://github.com/leanprover-community/mathlib4/commit/4b3bc835188889a809182c7d32c2cdff519b0046).